challenge/contact.py

```diff
@@ -1,188 +0,0 @@
-            # TODO: check if page really html?
-            # TODO: set focus on page first, before any key commands
-            scan.actions.perform()
-            #pdb.set_trace()
-            aa = '//a[@href]'
-            aas = self.scan.web.find_elements_by_xpath(aa)
-            href = [normalized(a.get_attribute('href')) for a in aas]
-            # links with domain
-            lsd = set([h for h in href if self.domain in h])
-            # relative links (or mailto)
-            rls = set([h for h in href if 'http' not in h])
-            self.mail = set([h[7:] for h in href if 'mailto' in h])
-            self.found = lsd|rls
-        except: # limit catches to HTTP or parsing exceptions?
-            self.found = set() # bad links or selenium exceptions on stale elements ?
-
-    def __iter__(self):
-        return self
-
-    def __next__(self):
-        
-         while (0 < len(self.found)):
-             link = self.found.pop() # str = type(link)
-             url = urlparse(normalized(link))
-             scheme = str(url.scheme)
-             if 0 == len(scheme) or (re.match('http',scheme) and re.search(self.domain+'$',str(url.netloc))):
-                 todo = urlunparse(url)
-                 return todo
-         raise StopIteration
-
-def headers(page):
-    ret = {'Content-Type':None,'ETag':None}
-    try:
-        with urllib.request.urlopen(page) as response:
-             ret = response.info()
-    except:
-        pass # ignore any errors, in particular 404
-    return ret
-
-logging.basicConfig()
-log = logging.getLogger('fe')
-
-def is_html(content_type):
-    if content_type:
-        for mime in ['text/html','application/xhtml+xml']:
-            if mime in content_type:
-                return True
-    return False
-
-def isdigit(c):
-    return c in '0123456789'
-
-if __name__ == '__main__':
-
-    # FIXME: see readme.md
-    hack = ''
-    if 'Linux' == platform.system():
-        hack = '/home/pi/.mozilla/firefox/m9f853mx.default'
-
-    domain = '<domain>'
-    usage = sys.argv[0] + ' [-v(erbose)|-d(ebug)] ' + domain
-    if 1 < len(sys.argv):
-        if '-t' == sys.argv[1]:
-            log.setLevel(logging.DEBUG)
-            contacts = Links(ui(hack),'http://jana.com/contact','jana.com')
-            log.debug(str(contacts))
-            for email in contacts.mail:
-                print(email)
-            contacts.scan.web.close()
-            exit(0)
-        elif '-v' == sys.argv[1]:
-            log.setLevel(logging.INFO)
-            domain = sys.argv[2]
-        elif '-d' == sys.argv[1]:
-            log.setLevel(logging.DEBUG)
-            domain = sys.argv[2]
-            pdb.set_trace()
-        else:
-            domain = sys.argv[1]
-    else:
-        print(usage)
-        exit(1)
-
-    with ui(hack) as scan:
-        """
-        #pdb.set_trace()
-        """
-        page = normalized('http://www.'+domain)
-        scheduled.add(page)
-        while(scheduled):
```

Comment:
>  think we can refactor it for clarity?

Of course, there is lots of room for improvement with the whole thing.  I generally don't like scrolling to see whole loops/functions.  I will look at this again.